### ISSUE TITLE
object condensation

### ISSUE THREAD

**Issue by vhewes** [nugraph/nugraph#50]
i just pushed a branch that does the following:
- applies some minor tweaks to the object condensation loss
- overhauls the instance decoder
- hooks the instance decoder into the model

i've updated the [training notebook](https://github.com/exatrkx/NuGraph/blob/6d16b1ecc3743904b8d2dc0ebac50095414b3757/notebooks/train.ipynb) to reflect these changes; if you run this notebook, you should be able to train a model with the instance decoder enabled. i ran this notebook and saw that the instance loss was getting smaller, but the first action item here is to train it and see if it continues to improve.

the readme for this repository has instructions on how to start tensorboard – once you've started a training instance, it should be possible to monitor training and see how it progresses. from there, if the loss curves look like it's training, then we'll need to implement a post-processing algorithm like DBSCAN to see how well it does at actually clustering.

the first action item for you is to check out this branch, set the notebook running on Heimdall, and then use tensorboard to monitor the training metrics. let me know if you have any questions on any of this – once you've run thru those steps successfully, we can talk about next steps.

### to-do list
- [x] Test higher-dimensional embedding
- [x] Build ObjCon embedding into message-passing loop
- [ ] Implement influencer network and compare

**Comment by vhewes** [comment]
@Marcelolq i just added a notebook for testing the object condensation loss function, as discussed. you can find it [here](https://github.com/exatrkx/NuGraph/blob/feature/object-condensation/notebooks/test-objcon.ipynb) on the `feature/object-condensation` branch!

**Comment by vhewes** [comment]
as a next step, we want to feed all three of the network heads through a final shared convolutional layer in order to generate the object condensation embedding. despite what i said yesterday, i think it may be possible to implement this entirely in the initializer function `__init__`, without needing to make any changes to the `forward` function. we can accomplish this by passing the same network module into different `nn.Sequential` blocks.

the current state of the `init` function is

https://github.com/exatrkx/NuGraph/blob/2910c1d19635ad4c36b0c18189cdc02510971ec5/nugraph/models/decoders.py#L303-L321

in order to make this change, we want to make the body of the function look something like
```python3
        # add in the 3D networks here
        net_3d_coords = nn.Linear(node_features, instance_features)
        net_3d filter = nn.Linear(node_features. 1)

        self.net_coords = nn.ModuleDict()
        self.net_filter = nn.ModuleDict()
        for p in planes:
            self.net_filter[p] = nn.Sequential(
                nn.Linear(node_features, node_features), # note that number of output features in this
                                                         # block is now node_features instead of 1
                net_3d_filter, # add in net_3d_filter as an additional layer here
                nn.Sigmoid(),
            )
            # the coordinate network is no longer a single layer, but two layers wrapped in nn.Sequential
            self.net_coords[p] = nn.Sequential(
                nn.Linear(node_features, node_features), # linear layer unique to each plane
                net_3d_coords, # shared linear layer across all planes
            )
        self.dfs = []
```
i've added comments to this function in all the places where it's changed, so try cross-referencing the two versions and implementing these changes, and then try running it. you can commit and push your changes as you go, and i can take a look and provide feedback. let me know if you have any questions!
